src/NLog/ILogMessageFormatter.cs

```diff
@@ -0,0 +1,122 @@
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using System.Globalization;
+using NLog.StructuredEvents;
+
+namespace NLog
+{
+    interface ILogMessageFormatter
+    {
+        string FormatMessage(LogEventInfo logEvent);
+        ILogMessageFormatter SetFormatProvider(IFormatProvider formatProvider);
+        IFormatProvider FormatProvider { get; }
+        bool HasProperties(LogEventInfo logEvent);
+    }
+
+    class LogMessageFormatter : ILogMessageFormatter
+    {
+        public static readonly LogMessageFormatter Default = new LogMessageFormatter();
+
+        public IFormatProvider FormatProvider { get; private set; }
+
+        public bool HasProperties(LogEventInfo _) { return false; }
+
+        public string FormatMessage(LogEventInfo logEvent)
+        {
+            return string.Format(FormatProvider ?? CultureInfo.CurrentCulture, logEvent.Message, logEvent.Parameters);
+        }
+
+        public ILogMessageFormatter SetFormatProvider(IFormatProvider formatProvider)
+        {
+            if (ReferenceEquals(this, Default) && !ReferenceEquals(formatProvider, null))
+            {
+                return new LogMessageFormatter() { FormatProvider = formatProvider };
+            }
+            else
+            {
+                FormatProvider = formatProvider;
+                return this;
+            }
+        }
+    }
+
+    class LogMessageTemplateFormatter : ILogMessageFormatter
+    {
+        public static readonly LogMessageTemplateFormatter Default = new LogMessageTemplateFormatter();
+
+        public IFormatProvider FormatProvider { get; private set; }
+
+        public bool HasProperties(LogEventInfo logEvent)
+        {
+            return logEvent.Message != null && logEvent.Parameters != null && logEvent.Parameters.Length > 0;
```

Comment:
`logEvent.Message != null ` is strange in `HasProperties`. We have to rename or change the behavoir